packages/effect/src/internal/trie.ts

```diff
@@ -0,0 +1,713 @@
+    new TrieIterator(self as TrieImpl<V>, (_, value) => value, (key) => key.startsWith(prefix))
+)
+
+/** @internal */
+export const entriesWithPrefix = dual<
+  (prefix: string) => <V>(self: TR.Trie<V>) => IterableIterator<[string, V]>,
+  <V>(self: TR.Trie<V>, prefix: string) => IterableIterator<[string, V]>
+>(
+  2,
+  <V>(self: TR.Trie<V>, prefix: string): IterableIterator<[string, V]> =>
+    new TrieIterator(self as TrieImpl<V>, (key, value) => [key, value], (key) => key.startsWith(prefix))
+)
+
+/** @internal */
+export const toEntriesWithPrefix = dual<
+  (prefix: string) => <V>(self: TR.Trie<V>) => Array<[string, V]>,
+  <V>(self: TR.Trie<V>, prefix: string) => Array<[string, V]>
+>(
+  2,
+  <V>(self: TR.Trie<V>, prefix: string): Array<[string, V]> => Array.from(entriesWithPrefix(self, prefix))
+)
+
+/** @internal */
+export const get = dual<
+  (key: string) => <V>(self: TR.Trie<V>) => Option.Option<V>,
+  <V>(self: TR.Trie<V>, key: string) => Option.Option<V>
+>(
+  2,
+  <V>(self: TR.Trie<V>, key: string) => {
+    let n: Node.Node<V> | undefined = (self as TrieImpl<V>)._root
+    if (n === undefined || key.length === 0) return Option.none()
+    let cIndex = 0
+    while (cIndex < key.length) {
+      const c = key[cIndex]
+      if (c > n.key) {
+        if (n.right === undefined) {
+          return Option.none()
+        } else {
+          n = n.right
+        }
+      } else if (c < n.key) {
+        if (n.left === undefined) {
+          return Option.none()
+        } else {
+          n = n.left
+        }
+      } else {
+        if (cIndex === key.length - 1) {
+          if (n.value !== undefined) {
+            return Option.some(n.value)
+          }
+        } else {
+          if (n.mid === undefined) {
+            return Option.none()
+          } else {
+            n = n.mid
+            cIndex += 1
+          }
+        }
+      }
+    }
+    return Option.none()
+  }
+)
+
+/** @internal */
+export const has = dual<
+  (key: string) => <V>(self: TR.Trie<V>) => boolean,
+  <V>(self: TR.Trie<V>, key: string) => boolean
+>(2, (self, key) => Option.isSome(get(self, key)))
+
+/** @internal */
+export const unsafeGet = dual<
+  (key: string) => <V>(self: TR.Trie<V>) => V,
+  <V>(self: TR.Trie<V>, key: string) => V
+>(2, (self, key) => {
+  const element = get(self, key)
+  if (Option.isNone(element)) {
+    throw new Error("Error: Expected trie to contain key")
+  }
+  return element.value
+})
+
+/** @internal */
+export const remove = dual<
+  (key: string) => <V>(self: TR.Trie<V>) => TR.Trie<V>,
+  <V>(self: TR.Trie<V>, key: string) => TR.Trie<V>
+>(
+  2,
+  <V>(self: TR.Trie<V>, key: string) => {
+    let n: Node.Node<V> | undefined = (self as TrieImpl<V>)._root
+    if (n === undefined || key.length === 0) return self
+
+    const count = n.count - 1
+    // -1:left | 0:mid | 1:right
+    const d_stack: Array<Ordering.Ordering> = []
+    const n_stack: Array<Node.Node<V>> = []
```

Comment:
Just regarding style: We don't use snake case anywhere else in the code base.

Reply:
I borrowed this naming from `RedBlackTree`. Should we update there too?

https://github.com/Effect-TS/effect/blob/8eec87e311ce55281a98517e6df0ef103b43e8a8/packages/effect/src/internal/redBlackTree.ts#L278-L279